File: LeetCode/0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.java
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if (head.next == null || right - left == 0) return head;
        
        ListNode root = new ListNode(0);
        root.next = head;
        ListNode start = root;
        
        for (int i = 0; i < left - 1; i++) start = start.next;
        
        ListNode end = start.next;
        
        for (int i = 0; i < right - left; i++) {
            ListNode next = start.next;
            start.next = end.next;
            end.next = end.next.next;
            start.next.next = next;
        }
        
        return root.next;
    }
}

File: LeetCode/0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.kt
class Solution {
    fun reverseBetween(head: ListNode?, left: Int, right: Int): ListNode? {
        if (head?.next == null || right - left == 0) return head
        
        val root = ListNode(0)
        root.next = head
        var start = root
        
        for (i in 0 until left - 1) start = start.next
        
        var end = start.next
        
        for (i in 0 until right - left) {
            val next = start.next
            start.next = end.next
            end.next = end.next.next
            start.next.next = next
        }
        
        return root.next
    }
}